Global styles controllers: extend with WP_REST_Posts_Controller and WP_REST_Revisions_Controller

ramonjd opened the pull request:
This change: 


- extends `WP_REST_Global_Styles_Revisions_Controller` with `WP_REST_Revisions_Controller`
- extends `WP_REST_Global_Styles_Controller` with `WP_REST_Posts_Controller`

As part of this work, the PR also:

- Removes shared public methods
- Updates schema to use parent and parent controller's properties. Unsets unnecessary properties.
- Updates relevant tests. Deletes dupes.

Trac ticket: https://core.trac.wordpress.org/ticket/60131

## Related PRs/discussions

- https://github.com/WordPress/gutenberg/pull/35801
- https://github.com/WordPress/wordpress-develop/pull/5655 (ensure this PR is rebased on top)
- https://github.com/WordPress/wordpress-develop/pull/5622 (some of the changes in this PR will be redundant, e.g., `get_item` but we'll still need the single revision route)
- https://github.com/WordPress/gutenberg/pull/55827#discussion_r1399856581

ramonjd committed:
Initial commit.
WP_REST_Global_Styles_Revisions_Controller extends WP_REST_Controller

ramonjd committed:
Adding title to schema.

ramonjd committed:
Remove permissions checks and favour parent's

ramonjd committed:
Remove duplicated tests

ramonjd committed:
Reverting post type registration changes as it assigns an autosave controller. we

ramonjd committed:
Regenerated fixture

ramonjd committed:
Revert introducing title in schema

ramonjd commented:
@adamsilverstein No huge rush, but if you get a chance to take this for a test drive, I'd be very grateful 🙇🏻

ramonjd commented:
I double checked and the failing E2E test is also failing on trunk:

```console
[[chromium] › gutenberg-plugin.test.js:25:6 › Gutenberg plugin › should activate: tests/e2e/specs/gutenberg-plugin.test.js#L12](https://github.com/WordPress/wordpress-develop/pull/5699/files#annotation_16570999825)
  1) [chromium] › gutenberg-plugin.test.js:25:6 › Gutenberg plugin › should activate ───────────────

    Retry #1 ───────────────────────────────────────────────────────────────────────────────────────
    SyntaxError: Unexpected token '<', "<br />
    <b>"... is not valid JSON

      10 | 	test.beforeAll( async ( { requestUtils } ) => {
      11 | 		// Install Gutenberg plugin if it's not yet installed.
    > 12 | 		const pluginsMap = await requestUtils.getPluginsMap();
         | 		                   ^
      13 | 		if ( ! pluginsMap.gutenberg ) {
      14 | 			await requestUtils.rest( {
      15 | 				method: 'POST',
```

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -334,6 +288,7 @@ public function prepare_item_for_response( $post, $request ) {
 	 * Retrieves the revision's schema, conforming to JSON Schema.
 	 *
 	 * @since 6.3.0
+	 * @since 6.5.0 Merged parent and parent controller schemata.
```
```suggestion
	 * @since 6.5.0 Merged parent and parent controller schema data.
```

ramonjd replied:
You bet! Thanks for the tip.

ramonjd committed:
Update src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

spacedmonkey commented on tests/phpunit/tests/rest-api/rest-global-styles-revisions-controller.php:
```diff
@@ -380,413 +380,6 @@ public function test_get_item_schema() {
 		$this->assertArrayHasKey( 'modified_gmt', $properties, 'Schema properties array has "modified_gmt" key.' );
 	}
 
-	/**
```
@ramonjd Removing tests doesn't seem like a good move. Did these tests run and pass before? If so, let's leave them.

ramonjd replied:
Yes they were all running and passing. I removed them because they were clones of the `WP_REST_Revisions_Controller` tests, so I thought it was duplication.

No worries though, I'll put them back in. 

Thanks for the feedback!

ramonjd replied:
Done!

spacedmonkey commented on tests/qunit/fixtures/wp-api-generated.js:
```diff
@@ -10035,10 +10035,58 @@ mockedApiResponse.Schema = {
                             "maximum": 100,
                             "required": false
                         },
+                        "search": {
+                            "description": "Limit results to those matching a string.",
+                            "type": "string",
+                            "required": false
+                        },
+                        "exclude": {
+                            "description": "Ensure result set excludes specific IDs.",
+                            "type": "array",
+                            "items": {
+                                "type": "integer"
+                            },
+                            "default": [],
+                            "required": false
+                        },
+                        "include": {
+                            "description": "Limit result set to specific IDs.",
+                            "type": "array",
+                            "items": {
+                                "type": "integer"
+                            },
+                            "default": [],
+                            "required": false
+                        },
```
This means no functionality that doesn't there before right? Have we tested?

ramonjd replied:
Oh, this is from the inheritance. Good point. 

I'll work on it. 🍺

ramonjd replied:
I've removed these inherited query params in a `get_collection_params` override - they are not currently supported for global styles revisions. If it makes sense to allow them, then maybe in a follow up.

👍🏻

spacedmonkey reviewed (changes requested):
Let's leave the unit tests and not delete them. 

One issue here is the revisions controller currently guesses that the parent controller is a instance or sub class of post controller. Means means it expects some classes and properties that might not exist in `WP_REST_Global_Styles_Controller`.

ramonjd committed:
Reinstating global styles revision controller tests that were duplicated from WP_REST_Revisions_Controller tests

ramonjd committed:
Excluding search, include and exclude from the collection query params

Excluding order and order by from the collection query params

Comment

spacedmonkey reviewed (commented):
My biggest worry about this `WP_REST_Global_Styles_Controller` does not extend the `WP_REST_Posts_Controller`. Future refactors would might meaning breaks here, as class methods / properties do not match. I feel like if we are going to make this change, we go the whole way and convert `WP_REST_Global_Styles_Controller` to extend `WP_REST_Posts_Controller`. Then we can use the work done in https://github.com/WordPress/wordpress-develop/commit/1f51e1f4f6c81238fffed706ff165795ea34d522, to regsiter custom post / revision controllers when registering the global post type. Then this manaul regsiteration can be removed here. 

https://github.com/WordPress/wordpress-develop/blob/43d2455dc68072fdd43c3c800cc8c32590f23cbe/src/wp-includes/rest-api.php#L326-L332

ramonjd commented:
> I feel like if we are going to make this change, we go the whole way and convert WP_REST_Global_Styles_Controller to extend WP_REST_Posts_Controller.

Thanks for the feedback. 

Looks like this point was raised by @TimothyBJacobs back when the endpoint was introduced: https://github.com/WordPress/gutenberg/pull/35801#pullrequestreview-788267865

> I think in this case, considering how custom the global styles API is, I think it is fine to extend the base controller instead of using WP_REST_Posts_Controller.

@TimothyBJacobs do you think this is still the case? The "custom" nature of global styles, I think, still stands.

I'm happy to take a stab at it if we think inheriting `WP_REST_Posts_Controller` will bring more benefits as described [above](https://github.com/WordPress/wordpress-develop/pull/5699#pullrequestreview-1803197096).

jonnynews commented:
> I'm happy to take a stab at it if we think inheriting `WP_REST_Posts_Controller` will bring more benefits as described [above](https://github.com/WordPress/wordpress-develop/pull/5699#pullrequestreview-1803197096).

The big difference is now https://github.com/WordPress/wordpress-develop/commit/1f51e1f4f6c81238fffed706ff165795ea34d522 is committed, we can make custom revisions endpoints. I think now we can have custom revisions, it register this as a post endpoint and revisions endpoint. 

Let's try this change in another PR, see if it makes sense.

ramonjd committed:
Now that https://github.com/WordPress/wordpress-develop/pull/5655 has landed, use inherited methods for getting individual items.

ramonjd commented:
> Let's try this change in another PR, see if it makes sense.

Started here. More stuff and tests to come 👍🏻 

- https://github.com/WordPress/wordpress-develop/pull/6008

ramonjd commented:
> The big difference is now https://github.com/WordPress/wordpress-develop/commit/1f51e1f4f6c81238fffed706ff165795ea34d522 is committed, we can make custom revisions endpoints. I think now we can have custom revisions, it register this as a post endpoint and revisions endpoint.

Works pretty well over at https://github.com/WordPress/wordpress-develop/pull/6008, but I'm wondering if registering endpoints should also be optional - e.g., Global styles should be able to opt out of autosaves routes

Perhaps explicitly setting `'${ autosave | revision }_rest_controller_class'` to `false` when registering the post?

Ora  a `supports_autosaves`, similar to `supports_revisions`? 🤷🏻 

I'm happy to get a PR up to test this out.

spacedmonkey commented:
Closing in favour of #6022

ramonjd commented:
> Closing in favour of https://github.com/WordPress/wordpress-develop/pull/6022

@spacedmonkey Just checking, this PR was working as a stand alone patch. 

Did you close with the intention to port these changes over to #6022? Otherwise I think we should reopen.

spacedmonkey commented:
> @spacedmonkey Just checking, this PR was working as a stand alone patch.

@ramonjd I thought we were aligned on https://github.com/WordPress/wordpress-develop/pull/6022? I feel like #6022 is good to commit.

ramonjd commented:
> @ramonjd I thought we were aligned on https://github.com/WordPress/wordpress-develop/pull/6022? I feel like https://github.com/WordPress/wordpress-develop/pull/6022 is good to commit.

Definitely, but just not sure why this PR needed to be closed?

I think maybe our wires just got crossed 😅 

This PR affects `WP_REST_Global_Styles_Revisions_Controller` (Revisions) not `WP_REST_Global_Styles_Controller`. I just assumed you were going to bundle them into #6022.

Happy to reapprove #6022 - it's good to go, but we still need this PR I think.

jonnynews commented:
@ramonjd To be clear, https://github.com/WordPress/wordpress-develop/pull/6022 needs to be part of this PR. Let's do this change in one commit.

ramonjd commented:
> To be clear, https://github.com/WordPress/wordpress-develop/pull/6022 needs to be part of this PR. Let's do this change in one commit.

Ah, got it. Thanks, for clearing that up. 

https://github.com/WordPress/wordpress-develop/pull/6022 seemed to work well on its own, but I can pull over the changes into this PR if that's your advice. 

Thanks again.

ramonjd committed:
Pulling across changes from https://github.com/WordPress/wordpress-develop/pull/6022

Updating generated API file

jonnynews commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -10,24 +10,20 @@
-		$this->namespace = 'wp/v2';
-		$this->rest_base = 'global-styles';
-		$this->post_type = 'wp_global_styles';
+	public function __construct( $post_type ) {
+		$this->post_type = $post_type;
+		$obj             = get_post_type_object( $post_type );
+		$this->rest_base = ! empty( $obj->rest_base ) ? $obj->rest_base : $obj->name;
+		$this->namespace = ! empty( $obj->rest_namespace ) ? $obj->rest_namespace : 'wp/v2';
 	}
```
I don't think we need to override the __construct here. I think the original one will work just fine.

ramonjd replied:
I think I left it out because `WP_REST_Posts_Controller` adds meta to the class object, `$this->meta = new WP_REST_Post_Meta_Fields( $this->post_type );`, but as you say, it works fine and as expected without the constructor. 

I'll remove. Cheers!

jonnynews commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -10,24 +10,20 @@
-	protected $post_type;
-
+class WP_REST_Global_Styles_Controller extends WP_REST_Posts_Controller {
 	/**
```
I think we should add 
```
protected $allow_batch = array( 'v1' => false );
```

To directly denote this endpoint does not support batching. This should be added to the register of the routes as well. See 
https://github.com/WordPress/wordpress-develop/blob/27087b71e0c109af68dbc0545893217fc80350db/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L91

This means it could be overridden later in a plugin.

ramonjd replied:
Good spotting!

jonnynews reviewed (changes requested):
I believe the contractor, get_item and update_item methods no longer need to be overridden in the global style controller. 

I also had a note about marking this endpoint as not support batching. 

I love this change, as it allows the gutenberg plugin to override these classes easier.

ramonjd committed:
Implementing latest round of feedback:

removing __construct, get_item and update_item in favour of the parent class methods

Add $allow_batch property, setting it to false

ramonjd commented:
> I believe the contractor, get_item and update_item methods no longer need to be overridden in the global style controller.

👍🏻 Done

What do you think about `get_post()` as well? 

The functions are the identical, the only difference being the error message, which seems fairly samey.

In [global styles controller](https://github.com/ramonjd/wordpress-develop/blob/dac8e0b0d6c098f82a44e624ed8abb2f19113577/src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php#L119-L119):

```php
		$error = new WP_Error(
			'rest_global_styles_not_found',
			__( 'No global styles config exist with that id.' ),
			array( 'status' => 404 )
		);

```

In [post controller](https://github.com/ramonjd/wordpress-develop/blob/dac8e0b0d6c098f82a44e624ed8abb2f19113577/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L455-L455):

```php
		$error = new WP_Error(
			'rest_post_invalid_id',
			__( 'Invalid post ID.' ),
			array( 'status' => 404 )
		);
```

spacedmonkey reviewed (approved):
Looks good to me

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -457,10 +382,10 @@ protected function prepare_links( $id ) {
 	 * @return array List of link relations.
 	 */
-	protected function get_available_actions() {
+	protected function get_available_actions( $post, $request ) {
```
Missing `@param` documentation for new variables.

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -35,12 +35,21 @@ class WP_REST_Global_Styles_Revisions_Controller extends WP_REST_Controller {
-		$this->rest_base        = 'revisions';
-		$this->parent_base      = 'global-styles';
-		$this->namespace        = 'wp/v2';
+	public function __construct( $parent_post_type ) {
```
Missing `@param` documentation for new variables.

jonnynews commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -457,10 +382,10 @@ protected function prepare_links( $id ) {
 	 *
```
```suggestion
	 * @since 6.5.0 Add $post and $request parameters. 
	 *
	 * @param WP_Post         $post    Post object.
	 * @param WP_REST_Request $request Request object.
```

jonnynews replied:
@mukeshpanchal27 ^

jonnynews commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-revisions-controller.php:
```diff
@@ -35,12 +35,21 @@ class WP_REST_Global_Styles_Revisions_Controller extends WP_REST_Controller {
 	 *
 	 * @since 6.3.0
+	 * @since 6.5.0 Extends class from WP_REST_Revisions_Controller.
 	 */
```
```suggestion
	 *
	 * @param string $parent_post_type Post type of the parent.
	 */
```

spacedmonkey committed:
Apply suggestions from code review

Co-authored-by: Jonny Harris <148061917+jonnynews@users.noreply.github.com>

mukeshpanchal27 commented on src/wp-includes/rest-api/endpoints/class-wp-rest-global-styles-controller.php:
```diff
@@ -454,13 +379,16 @@ protected function prepare_links( $id ) {
 	 *
 	 * @since 5.9.0
 	 * @since 6.2.0 Added 'edit-css' action.
+	 * @since 6.5.0 Add $post and $request parameters. 
```
```suggestion
	 * @since 6.5.0 Added $post and $request parameters.
```

spacedmonkey commented:
Committed.
